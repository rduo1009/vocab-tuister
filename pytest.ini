[pytest]
addopts = 
    --strict-markers
    --ignore=src/_vendor
    -n4

markers =
    manual: marks tests as manually checked