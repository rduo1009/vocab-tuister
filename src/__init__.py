"""A wrapper for the Python packages used by ``vocab-tester``."""

from __future__ import annotations

import os as _os

_seed: str | None = _os.getenv("VOCAB_TUISTER_RANDOM_SEED")
if _seed is not None:
    import random
    import sys as _sys
    import types as _types

    try:
        _seed_value: int = int(_seed)
    except ValueError as e:
        raise ValueError(
            f"Invalid seed value: {_seed}. Must be an integer."
        ) from e

    _custom_random = _types.ModuleType("random")
    _custom_random_class = random.Random(_seed_value)

    _custom_random.Random = random.Random  # type: ignore[attr-defined]
    _custom_random.SystemRandom = random.SystemRandom  # type: ignore[attr-defined]

    _custom_random.random = _custom_random_class.random  # type: ignore[attr-defined]
    _custom_random.uniform = _custom_random_class.uniform  # type: ignore[attr-defined]

    _custom_random.randint = _custom_random_class.randint  # type: ignore[attr-defined]
    _custom_random.randrange = _custom_random_class.randrange  # type: ignore[attr-defined]

    _custom_random.choice = _custom_random_class.choice  # type: ignore[attr-defined]
    _custom_random.sample = _custom_random_class.sample  # type: ignore[attr-defined]
    _custom_random.shuffle = _custom_random_class.shuffle  # type: ignore[attr-defined]

    _sys.modules["random"] = _custom_random


import dunamai as _dunamai

from . import core, server

try:
    __version__ = _dunamai.get_version(
        "src", third_choice=_dunamai.Version.from_any_vcs
    ).serialize()
except RuntimeError:
    import sys as _sys

    # Frozen with PyInstaller
    if getattr(_sys, "frozen", False) and hasattr(
        _sys, "_MEIPASS"
    ):  # pragma: no cover
        from pathlib import Path

        version_path = Path(__file__).parent.parent / "__version__.txt"

        __version__ = version_path.read_text().strip()
    else:
        raise

__author__ = "rduo1009"
__copyright__ = "2024, rduo1009"
__license__ = "MIT"
__all__ = ["core", "server"]
