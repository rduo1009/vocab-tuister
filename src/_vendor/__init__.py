from . import aenum, typeddict_validator