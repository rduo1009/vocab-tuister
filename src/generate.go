package main

//go:generate go run structs_generator.go
//go:generate python3 scripts/create_settings_json.py
//go:generate gojsonstruct --typename SessionConfig --package-name pkg --file-header "// Code generated by gojsonstruct; DO NOT EDIT." -o client/pkg/sessionconfig.go scripts/json_output/Settings_sample.json

//go:generate mkunion watch -g client/pkg/questions/questions.go
//go:generate mkunion
//go:generate patch client/pkg/questions/types_reg_gen.go client/pkg/questions/types_reg_gen_fix.diff

//go:generate go tool enumer -type=Number,Tense,Voice,Mood,Case,Gender,Degree -transform=snake -output client/pkg/enums/ending_components_enumer_gen.go client/pkg/enums/ending_components.go
