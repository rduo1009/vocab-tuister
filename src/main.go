package main

import "github.com/rduo1009/vocab-tuister/src/client/cmd"

func main() {
	cmd.Execute()
}
