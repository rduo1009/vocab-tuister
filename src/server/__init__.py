"""Contains the server for vocab-tester."""

from . import app, exceptions

__all__ = ["app", "exceptions"]
