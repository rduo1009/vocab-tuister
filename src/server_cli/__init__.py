"""Contains the cli that runs the server."""
