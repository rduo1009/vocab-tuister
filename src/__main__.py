"""The CLI that runs the server."""

# ruff: noqa: TCH002

from __future__ import annotations

import logging
import sys
import warnings
from typing import Annotated

from cyclopts import App, Parameter
from cyclopts.types import UInt16

from src import __version__
from src.utils.logger import (
    CustomHandler,
    custom_formatwarning,
    log_uncaught_exceptions,
)

from .server.app import main as server_main
from .server.app import main_dev as server_main_dev

# Initialise logger
logger = logging.getLogger()
logging.captureWarnings(capture=True)

# Set custom formatting for warnings and exceptions
warnings.formatwarning = custom_formatwarning
sys.excepthook = log_uncaught_exceptions  # type: ignore[assignment]

# Set custom handler for logs
for handler in logger.handlers.copy():
    logger.removeHandler(handler)
handler = CustomHandler()


logger.addHandler(handler)
cli = App(
    version=__version__,  # for some reason this is needed
    default_parameter=Parameter(negative=()),
)


def _set_verbosity(verbosity: int) -> None:
    verbosity_map = {
        -1: logging.CRITICAL,  # quiet mode
        0: logging.ERROR,
        1: logging.WARNING,
        2: logging.INFO,
        3: logging.DEBUG,
    }

    logger.setLevel(verbosity_map.get(verbosity, logging.DEBUG))
    handler.setLevel(verbosity_map.get(verbosity, logging.DEBUG))


@cli.default
def vocab_tuister_server(
    *,
    port: Annotated[UInt16, Parameter(name=["--port", "-p"])] = 5000,
    verbose: Annotated[
        tuple[bool, ...], Parameter(name=["--verbose", "-v"])
    ] = (False,),
    quiet: bool = False,
    dev: bool = False,
    debug: bool = False,
) -> None:
    """Start the vocab-tuister server.

    Parameters
    ----------
    port : int
        The port to run the server on.
    verbose : tuple[bool, ...]
        How verbose to make the output (maximum 3). Note that additional
        verbosity flags must be added like ``-v -v -v``, due to a bug with
        Cyclopts.
    quiet : bool
        Whether to make the output quiet (only critical errors). Overrides
        ``--verbose``.
    dev : bool
        Use the Flask development server instead of the production server.
        Should not be used usually.
    debug : bool
        Run in debug mode. Implies ``--dev``.
        Should not be used usually.
    """
    _set_verbosity(-1 if quiet else sum(verbose))

    if dev or debug:
        server_main_dev(port, debug=debug)
    else:
        server_main(port)


if __name__ == "__main__":
    cli()
