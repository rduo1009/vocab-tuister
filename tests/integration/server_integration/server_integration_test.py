import os
import sys

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))

import contextlib
import json
import os
import threading
from pathlib import Path
from time import sleep

import pytest
import requests
from src.__main__ import cli


def run_cli(port):
    with contextlib.suppress(KeyboardInterrupt):
        cli(["-v", "-v", "-v", "-p", str(port)])


def setup_tests(monkeypatch, port, vocab_file_info, session_config_info):
    monkeypatch.setattr("src.server.app.vocab_list", None)

    server_url = f"http://127.0.0.1:{port}"

    vocab_file = Path(__file__).parent / "testdata" / f"test-{vocab_file_info}-list.txt"
    vocab_list = vocab_file.read_text(encoding="utf-8")

    session_config_file = Path(__file__).parent / "testdata" / f"test-{session_config_info}-config.json"
    session_config = json.loads(session_config_file.read_text(encoding="utf-8"))

    cli_process = threading.Thread(target=run_cli, args=(port,), daemon=True)
    cli_process.start()

    return server_url, vocab_list, session_config, cli_process


@pytest.mark.integration
def test_cli_normal(snapshot, monkeypatch):
    server_url, vocab_list, session_config, cli_process = setup_tests(monkeypatch, 5500, "regular", "regular")

    try:
        sleep(5)

        vocab_response = requests.post(f"{server_url}/send-vocab", data=vocab_list, timeout=5)
        assert vocab_response.status_code == 200
        assert vocab_response.text == snapshot

        session_response = requests.post(f"{server_url}/session", json=session_config, timeout=5)
        assert session_response.status_code == 200
        assert session_response.json() == snapshot

    finally:
        sleep(5)

        cli_process.join(timeout=10)


@pytest.mark.integration
def test_cli_error_list(snapshot, monkeypatch):
    server_url, vocab_list, session_config, cli_process = setup_tests(monkeypatch, 5501, "error", "regular")

    try:
        sleep(5)

        vocab_response = requests.post(f"{server_url}/send-vocab", data=vocab_list, timeout=5)
        assert vocab_response.status_code == 400
        assert vocab_response.text == snapshot

        try:
            session_response = requests.post(f"{server_url}/session", json=session_config, timeout=5)
            session_response.raise_for_status()

            pytest.fail("Expected an error but request succeeded.")

        except requests.exceptions.HTTPError:
            if session_response.status_code == 400:
                assert session_response.status_code == 400
                assert session_response.text == snapshot
            else:
                raise

        except requests.exceptions.RequestException:
            raise

    finally:
        sleep(5)

        cli_process.join(timeout=10)


@pytest.mark.integration
def test_cli_error_config(snapshot, monkeypatch):
    server_url, vocab_list, session_config, cli_process = setup_tests(monkeypatch, 5502, "regular", "error")
    try:
        sleep(5)

        vocab_response = requests.post(f"{server_url}/send-vocab", data=vocab_list, timeout=5)
        assert vocab_response.status_code == 200
        assert vocab_response.text == snapshot

        try:
            session_response = requests.post(f"{server_url}/session", json=session_config, timeout=5)
            session_response.raise_for_status()

            pytest.fail("Expected an error but request succeeded.")

        except requests.exceptions.HTTPError:
            if session_response.status_code == 400:
                assert session_response.status_code == 400
                assert session_response.text == snapshot
            else:
                raise

        except requests.exceptions.RequestException:
            raise

    finally:
        sleep(5)

        cli_process.join(timeout=10)
