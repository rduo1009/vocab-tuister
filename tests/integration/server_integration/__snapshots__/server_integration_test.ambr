# serializer version: 1
# name: test_cli_error_extra_config
  'Vocab list received.'
# ---
# name: test_cli_error_extra_config.1
  "Bad request: 400 Bad Request: Required settings are missing: 'exclude-deponents', 'exclude-participle-future-active', 'exclude-verb-future-passive-indicative', 'exclude-verb-future-perfect-passive-indicative', 'exclude-verb-imperfect-passive-indicative', 'exclude-verb-perfect-active-subjunctive', 'exclude-verb-perfect-passive-indicative', 'exclude-verb-pluperfect-passive-indicative', 'exclude-verb-present-active-subjunctive', 'exclude-verb-present-passive-indicative', 'exclude-verb-present-passive-infinitive'. (InvalidSettingsError)"
# ---
# name: test_cli_error_list
  "Bad request: 400 Bad Request: InvalidVocabFileFormatError: Invalid part of speech: 'Cause an error'"
# ---
# name: test_cli_error_list.1
  'Bad request: 400 Bad Request: Vocab list has not been provided.'
# ---
# name: test_cli_error_missing1_config
  'Vocab list received.'
# ---
# name: test_cli_error_missing1_config.1
  "Bad request: 400 Bad Request: Required settings are missing: 'number-of-questions'. (InvalidSettingsError)"
# ---
# name: test_cli_error_missing2_config
  'Vocab list received.'
# ---
# name: test_cli_error_missing2_config.1
  "Bad request: 400 Bad Request: Required settings are missing: 'exclude-deponents', 'exclude-participle-future-active', 'exclude-verb-future-passive-indicative', 'exclude-verb-future-perfect-passive-indicative', 'exclude-verb-imperfect-passive-indicative', 'exclude-verb-perfect-active-subjunctive', 'exclude-verb-perfect-passive-indicative', 'exclude-verb-pluperfect-passive-indicative', 'exclude-verb-present-active-subjunctive', 'exclude-verb-present-passive-indicative', 'exclude-verb-present-passive-infinitive', 'number-multiplechoice-options'. (InvalidSettingsError)"
# ---
# name: test_cli_error_type1_config
  'Vocab list received.'
# ---
# name: test_cli_error_type1_config.1
  "Bad request: 400 Bad Request: Required settings are missing: 'exclude-deponents', 'exclude-participle-future-active', 'exclude-verb-future-passive-indicative', 'exclude-verb-future-perfect-passive-indicative', 'exclude-verb-imperfect-passive-indicative', 'exclude-verb-perfect-active-subjunctive', 'exclude-verb-perfect-passive-indicative', 'exclude-verb-pluperfect-passive-indicative', 'exclude-verb-present-active-subjunctive', 'exclude-verb-present-passive-indicative', 'exclude-verb-present-passive-infinitive'. (InvalidSettingsError)"
# ---
# name: test_cli_error_type2_config
  'Vocab list received.'
# ---
# name: test_cli_error_type2_config.1
  "Bad request: 400 Bad Request: Invalid settings: 'number-of-questions' must be an integer (got type str). (InvalidSettingsError)"
# ---
# name: test_cli_normal
  'Vocab list received.'
# ---
# name: test_cli_normal.1
  list([
    dict({
      'MultipleChoiceEngToLatQuestion': dict({
        'answer': 'hic',
        'choices': list([
          'acer',
          'hic',
          'laetus',
        ]),
        'prompt': 'this',
      }),
      'question_type': 'MultipleChoiceEngToLatQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'dative singular',
          'genitive singular',
          'nominative plural',
          'vocative plural',
        ]),
        'dictionary_entry': 'farmer: agricola, agricolae, (m)',
        'main_answer': 'nominative plural',
        'prompt': 'agricolae',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'TypeInEngToLatQuestion': dict({
        'answers': list([
          'canem',
          'canis',
        ]),
        'main_answer': 'canem',
        'prompt': 'dog',
      }),
      'question_type': 'TypeInEngToLatQuestion',
    }),
    dict({
      'MultipleChoiceEngToLatQuestion': dict({
        'answer': 'ego',
        'choices': list([
          'levis',
          'puer',
          'ego',
        ]),
        'prompt': 'I',
      }),
      'question_type': 'MultipleChoiceEngToLatQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'comparative ablative singular feminine',
          'comparative ablative singular masculine',
          'comparative ablative singular neuter',
        ]),
        'dictionary_entry': 'large: ingens, ingentis, (3-1)',
        'main_answer': 'comparative ablative singular masculine',
        'prompt': 'ingentiore',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'TypeInLatToEngQuestion': dict({
        'answers': list([
          'by means of the names',
          'by the names',
          'for names',
          'for the names',
          'names',
          'to names',
          'to the names',
          'with the names',
        ]),
        'main_answer': 'for the names',
        'prompt': 'nominibus',
      }),
      'question_type': 'TypeInLatToEngQuestion',
    }),
    dict({
      'MultipleChoiceLatToEngQuestion': dict({
        'answer': 'good',
        'choices': list([
          'say',
          'good',
          'large',
        ]),
        'prompt': 'bonus',
      }),
      'question_type': 'MultipleChoiceLatToEngQuestion',
    }),
    dict({
      'PrincipalPartsQuestion': dict({
        'principal_parts': list([
          'bonus',
          'bona',
          'bonum',
        ]),
        'prompt': 'bonus',
      }),
      'question_type': 'PrincipalPartsQuestion',
    }),
    dict({
      'MultipleChoiceEngToLatQuestion': dict({
        'answer': 'levis',
        'choices': list([
          'ille',
          'audio',
          'levis',
        ]),
        'prompt': 'light',
      }),
      'question_type': 'MultipleChoiceEngToLatQuestion',
    }),
    dict({
      'ParseWordCompToLatQuestion': dict({
        'answers': list([
          'agricolas',
        ]),
        'components': 'accusative plural',
        'main_answer': 'agricolas',
        'prompt': 'farmer: agricola, agricolae, (m)',
      }),
      'question_type': 'ParseWordCompToLatQuestion',
    }),
    dict({
      'TypeInLatToEngQuestion': dict({
        'answers': list([
          'large',
        ]),
        'main_answer': 'large',
        'prompt': 'ingentia',
      }),
      'question_type': 'TypeInLatToEngQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'positive accusative plural feminine',
          'positive accusative plural masculine',
          'positive nominative plural feminine',
          'positive nominative plural masculine',
          'positive vocative plural feminine',
          'positive vocative plural masculine',
        ]),
        'dictionary_entry': 'keen: acer, acris, acre, (3-3)',
        'main_answer': 'positive nominative plural masculine',
        'prompt': 'acres',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'ParseWordCompToLatQuestion': dict({
        'answers': list([
          'horum',
        ]),
        'components': 'genitive plural neuter',
        'main_answer': 'horum',
        'prompt': 'this: hic, haec, hoc',
      }),
      'question_type': 'ParseWordCompToLatQuestion',
    }),
    dict({
      'MultipleChoiceLatToEngQuestion': dict({
        'answer': 'into',
        'choices': list([
          'keen',
          'boy',
          'into',
        ]),
        'prompt': 'in',
      }),
      'question_type': 'MultipleChoiceLatToEngQuestion',
    }),
    dict({
      'ParseWordCompToLatQuestion': dict({
        'answers': list([
          'laetissima',
        ]),
        'components': 'superlative accusative plural neuter',
        'main_answer': 'laetissima',
        'prompt': 'happy: laetus, laeta, laetum, (2-1-2)',
      }),
      'question_type': 'ParseWordCompToLatQuestion',
    }),
    dict({
      'TypeInEngToLatQuestion': dict({
        'answers': list([
          'illius',
        ]),
        'main_answer': 'illius',
        'prompt': 'of that',
      }),
      'question_type': 'TypeInEngToLatQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'perfect active indicative singular 1st person',
        ]),
        'dictionary_entry': 'say: inquam',
        'main_answer': 'perfect active indicative singular 1st person',
        'prompt': 'inquii',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'PrincipalPartsQuestion': dict({
        'principal_parts': list([
          'ingens',
          'ingentis',
        ]),
        'prompt': 'ingens',
      }),
      'question_type': 'PrincipalPartsQuestion',
    }),
    dict({
      'MultipleChoiceEngToLatQuestion': dict({
        'answer': 'inquam',
        'choices': list([
          'ingredior',
          'laetus',
          'inquam',
        ]),
        'prompt': 'say',
      }),
      'question_type': 'MultipleChoiceEngToLatQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'comparative ablative singular feminine',
          'comparative ablative singular masculine',
          'comparative ablative singular neuter',
        ]),
        'dictionary_entry': 'large: ingens, ingentis, (3-1)',
        'main_answer': 'comparative ablative singular masculine',
        'prompt': 'ingentiore',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'MultipleChoiceLatToEngQuestion': dict({
        'answer': 'good',
        'choices': list([
          'say',
          'from',
          'good',
        ]),
        'prompt': 'bonus',
      }),
      'question_type': 'MultipleChoiceLatToEngQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'dative singular',
        ]),
        'dictionary_entry': 'dog: canis, canis, (m)',
        'main_answer': 'dative singular',
        'prompt': 'cani',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'TypeInEngToLatQuestion': dict({
        'answers': list([
          'puellis',
        ]),
        'main_answer': 'puellis',
        'prompt': 'for the girls',
      }),
      'question_type': 'TypeInEngToLatQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'ablative singular',
          'nominative singular',
          'vocative singular',
        ]),
        'dictionary_entry': 'girl: puella, puellae, (f)',
        'main_answer': 'nominative singular',
        'prompt': 'puella',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'genitive singular',
        ]),
        'dictionary_entry': 'I: ego, (irregular)',
        'main_answer': 'genitive singular',
        'prompt': 'mei',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'PrincipalPartsQuestion': dict({
        'principal_parts': list([
          'audio',
          'audire',
          'audivi',
          'auditus',
        ]),
        'prompt': 'audio',
      }),
      'question_type': 'PrincipalPartsQuestion',
    }),
    dict({
      'ParseWordCompToLatQuestion': dict({
        'answers': list([
          'capiendi',
        ]),
        'components': 'gerundive masculine genitive singular',
        'main_answer': 'capiendi',
        'prompt': 'take: capio, capere, cepi, captus',
      }),
      'question_type': 'ParseWordCompToLatQuestion',
    }),
    dict({
      'MultipleChoiceEngToLatQuestion': dict({
        'answer': 'puer',
        'choices': list([
          'laetus',
          'puer',
          'nomen',
        ]),
        'prompt': 'boy',
      }),
      'question_type': 'MultipleChoiceEngToLatQuestion',
    }),
    dict({
      'TypeInLatToEngQuestion': dict({
        'answers': list([
          'lighter',
          'more light',
        ]),
        'main_answer': 'lighter',
        'prompt': 'levioris',
      }),
      'question_type': 'TypeInLatToEngQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'genitive singular feminine',
          'genitive singular masculine',
          'genitive singular neuter',
        ]),
        'dictionary_entry': 'that: ille, illa, illud',
        'main_answer': 'genitive singular masculine',
        'prompt': 'illius',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'ablative plural',
          'dative plural',
        ]),
        'dictionary_entry': 'girl: puella, puellae, (f)',
        'main_answer': 'dative plural',
        'prompt': 'puellis',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'positive ablative singular feminine',
          'positive accusative plural neuter',
          'positive nominative plural neuter',
          'positive nominative singular feminine',
          'positive vocative plural neuter',
          'positive vocative singular feminine',
        ]),
        'dictionary_entry': 'good: bonus, bona, bonum, (2-1-2)',
        'main_answer': 'positive nominative singular feminine',
        'prompt': 'bona',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'PrincipalPartsQuestion': dict({
        'principal_parts': list([
          'acer',
          'acris',
          'acre',
        ]),
        'prompt': 'acer',
      }),
      'question_type': 'PrincipalPartsQuestion',
    }),
    dict({
      'ParseWordCompToLatQuestion': dict({
        'answers': list([
          'harum',
        ]),
        'components': 'genitive plural feminine',
        'main_answer': 'harum',
        'prompt': 'this: hic, haec, hoc',
      }),
      'question_type': 'ParseWordCompToLatQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'superlative ablative singular feminine',
          'superlative accusative plural neuter',
          'superlative nominative plural neuter',
          'superlative nominative singular feminine',
          'superlative vocative plural neuter',
          'superlative vocative singular feminine',
        ]),
        'dictionary_entry': 'light: levis, leve, (3-2)',
        'main_answer': 'superlative nominative singular feminine',
        'prompt': 'levissima',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'PrincipalPartsQuestion': dict({
        'principal_parts': list([
          'puella',
          'puellae',
        ]),
        'prompt': 'puella',
      }),
      'question_type': 'PrincipalPartsQuestion',
    }),
    dict({
      'MultipleChoiceEngToLatQuestion': dict({
        'answer': 'bonus',
        'choices': list([
          'capio',
          'acer',
          'bonus',
        ]),
        'prompt': 'good',
      }),
      'question_type': 'MultipleChoiceEngToLatQuestion',
    }),
    dict({
      'MultipleChoiceEngToLatQuestion': dict({
        'answer': 'e',
        'choices': list([
          'capio',
          'ingens',
          'e',
        ]),
        'prompt': 'from',
      }),
      'question_type': 'MultipleChoiceEngToLatQuestion',
    }),
    dict({
      'PrincipalPartsQuestion': dict({
        'principal_parts': list([
          'ingredior',
          'ingredi',
          'ingressus sum',
          'ingressus',
        ]),
        'prompt': 'ingredior',
      }),
      'question_type': 'PrincipalPartsQuestion',
    }),
    dict({
      'TypeInLatToEngQuestion': dict({
        'answers': list([
          'best',
          'extremely good',
          'most good',
          'quite good',
          'rather good',
          'too good',
          'very good',
        ]),
        'main_answer': 'best',
        'prompt': 'optima',
      }),
      'question_type': 'TypeInLatToEngQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'present active imperative singular 2nd person',
        ]),
        'dictionary_entry': 'say: inquam',
        'main_answer': 'present active imperative singular 2nd person',
        'prompt': 'inque',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'TypeInLatToEngQuestion': dict({
        'answers': list([
          'keener',
          'more keen',
        ]),
        'main_answer': 'keener',
        'prompt': 'acriori',
      }),
      'question_type': 'TypeInLatToEngQuestion',
    }),
    dict({
      'TypeInEngToLatQuestion': dict({
        'answers': list([
          'mihi',
        ]),
        'main_answer': 'mihi',
        'prompt': 'to me',
      }),
      'question_type': 'TypeInEngToLatQuestion',
    }),
    dict({
      'ParseWordCompToLatQuestion': dict({
        'answers': list([
          'acerrimis',
        ]),
        'components': 'superlative ablative plural masculine',
        'main_answer': 'acerrimis',
        'prompt': 'keen: acer, acris, acre, (3-3)',
      }),
      'question_type': 'ParseWordCompToLatQuestion',
    }),
    dict({
      'TypeInEngToLatQuestion': dict({
        'answers': list([
          'inquimus',
        ]),
        'main_answer': 'inquimus',
        'prompt': 'we say',
      }),
      'question_type': 'TypeInEngToLatQuestion',
    }),
    dict({
      'MultipleChoiceEngToLatQuestion': dict({
        'answer': 'ille',
        'choices': list([
          'agricola',
          'laetus',
          'ille',
        ]),
        'prompt': 'that',
      }),
      'question_type': 'MultipleChoiceEngToLatQuestion',
    }),
    dict({
      'ParseWordLatToCompQuestion': dict({
        'answers': list([
          'genitive singular',
        ]),
        'dictionary_entry': 'name: nomen, nominis, (n)',
        'main_answer': 'genitive singular',
        'prompt': 'nominis',
      }),
      'question_type': 'ParseWordLatToCompQuestion',
    }),
    dict({
      'MultipleChoiceEngToLatQuestion': dict({
        'answer': 'nomen',
        'choices': list([
          'hic',
          'capio',
          'nomen',
        ]),
        'prompt': 'name',
      }),
      'question_type': 'MultipleChoiceEngToLatQuestion',
    }),
    dict({
      'TypeInEngToLatQuestion': dict({
        'answers': list([
          'puellae',
          'puellas',
        ]),
        'main_answer': 'puellae',
        'prompt': 'girls',
      }),
      'question_type': 'TypeInEngToLatQuestion',
    }),
    dict({
      'PrincipalPartsQuestion': dict({
        'principal_parts': list([
          'ingens',
          'ingentis',
        ]),
        'prompt': 'ingens',
      }),
      'question_type': 'PrincipalPartsQuestion',
    }),
  ])
# ---
